### PR TITLE
[Modules] Added Alias

### PR DESCRIPTION
**Pull request by AlexanderSehr**
# Change

- Alias: Added module

TODO: 
- Missing logic to move removed subsription into 'Decomissioned' management group

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update (Wiki)

### CODE REVIEW

**Review by MrMCake — COMMENTED** [review]
Missing logic to move removed subsription into Decomissioned management group

File: arm/Microsoft.Subscription/aliases/deploy.bicep [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+param billingAccount string
+
+@description('Required. Alias to assign to the subscription')
+param subscriptionAlias string
+
+@description('Required. Display name for the subscription')
+param subscriptionDisplayName string
+
+@description('Required. Workload type for the subscription')
+param subscriptionWorkload string
```
Comment (MariusStorhaug) [inline arm/Microsoft.Subscription/aliases/deploy.bicep:16]:
```suggestion
@allowed([
  'Production'
  'DevTest'
])
@description('Optional. The workload type of the subscription.')
param subscriptionWorkload string = 'Production'
```

File: arm/Microsoft.Subscription/aliases/deploy.bicep [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+resource alias 'Microsoft.Subscription/aliases@2021-10-01' = {
+  scope: tenant()
+  name: subscriptionAlias
+  properties: {
+    workload: subscriptionWorkload
+    displayName: empty(subscriptionId) ? subscriptionDisplayName : null
+    billingScope: empty(subscriptionId) ? tenantResourceId('Microsoft.Billing/billingAccounts/enrollmentAccounts', billingAccount, enrollmentAccount) : null
+    additionalProperties: {
+      managementGroupId: !empty(managementGroupId) ? tenantResourceId('Microsoft.Management/managementGroups', managementGroupId) : null
```
Comment (MariusStorhaug) [inline arm/Microsoft.Subscription/aliases/deploy.bicep:41]: Think we need to choose an approach. Either this or the subscriptionPlacement module

File: arm/Microsoft.Subscription/aliases/readme.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+| :-- | :-- | :-- | :-- | :-- |
+| `billingAccount` | string |  |  | Required. BillingAccount used for subscription billing |
+| `enrollmentAccount` | string |  |  | Required. EnrollmentAccount used for subscription billing |
+| `managementGroupId` | string |  |  | Optional. The ID of the management group to deploy into. If not provided the subscription is deployed into the root management group |
+| `subscriptionAlias` | string |  |  | Required. Alias to assign to the subscription |
+| `subscriptionDisplayName` | string |  |  | Required. Display name for the subscription |
+| `subscriptionId` | string |  |  | Optional. This parameter can be used to create alias for existing subscription Id |
+| `subscriptionOwnerId` | string |  |  | Optional. Owner Id of the subscription |
+| `subscriptionTenantId` | string |  |  | Optional. Tenant Id of the subscription |
+| `subscriptionWorkload` | string |  |  | Required. Workload type for the subscription |
```
Comment (MariusStorhaug) [inline arm/Microsoft.Subscription/aliases/readme.md:24]: Need an update here if you go with the suggestion on default value.
```suggestion
| `subscriptionWorkload` | string | `Production` | `[Production, DevTest]` | Optional. The workload type of the subscription. |
```

**Review by MariusStorhaug — CHANGES_REQUESTED** [review]

File: arm/Microsoft.Subscription/aliases/deploy.bicep (hunk 2 above)
Comment (MrMCake) [reply arm/Microsoft.Subscription/aliases/deploy.bicep:45]: Yes and no. This property did not work for me, yet it is documented that it should. Hence I had to add the placement module regardless. I guess further tests will be necessary.

File: arm/Microsoft.Subscription/aliases/deploy.bicep [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+param subscriptionWorkload string = 'Production'
+
+@description('Optional. Owner Id of the subscription')
+param subscriptionOwnerId string = ''
+
+@description('Optional. Tenant Id of the subscription')
+param subscriptionTenantId string = ''
+
+@description('Optional. This parameter can be used to create alias for existing subscription Id')
+param subscriptionId string = ''
```
Comment (MariusStorhaug) [inline arm/Microsoft.Subscription/aliases/deploy.bicep:29]: As this module is about a subscription, would it make sense to make this more object oriented, not mentioning `subscription` in the parameters which specifically refer to the resource/object its managing?

File: arm/Microsoft.Subscription/aliases/deploy.bicep [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+module subscriptionPlacement '.bicep/mgmtGroup.bicep' = {
+  scope: managementGroup(managementGroupId)
+  name: '${uniqueString(deployment().name)}-placement'
+  params: {
+    managementGroupId: managementGroupId
+    subscriptionId: alias.properties.subscriptionId
+  }
+}
+
```
Comment (MariusStorhaug) [inline arm/Microsoft.Subscription/aliases/deploy.bicep:62]: missing ability to handle rbac within the subscription definition. Thinking we have the same capability on management group and resource group, and we should have it here too.

File: arm/Microsoft.Subscription/aliases/deploy.bicep [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+  name: subscriptionAlias
+  properties: {
+    workload: subscriptionWorkload
+    displayName: empty(subscriptionId) ? subscriptionDisplayName : null
+    billingScope: empty(subscriptionId) ? tenantResourceId('Microsoft.Billing/billingAccounts/enrollmentAccounts', billingAccount, enrollmentAccount) : null
+    additionalProperties: {
+      managementGroupId: !empty(managementGroupId) ? tenantResourceId('Microsoft.Management/managementGroups', managementGroupId) : null
+      subscriptionOwnerId: !empty(subscriptionOwnerId) ? subscriptionOwnerId : null
+      subscriptionTenantId: !empty(subscriptionTenantId) ? subscriptionTenantId : null
+      tags: tags
```
Comment (MariusStorhaug) [inline arm/Microsoft.Subscription/aliases/deploy.bicep:48]: Early tests showed me that this `tags` property only sets the first time and is not idempotent. Please verify that when we get that far.

File: arm/Microsoft.Subscription/aliases/deploy.bicep (hunk 5 above)
Comment (MrMCake) [reply arm/Microsoft.Subscription/aliases/deploy.bicep:62]: True. Just didn't add it as I wasn't sure if the alias creates the subscription fast enough for Rbac

File: arm/Microsoft.Subscription/aliases/deploy.bicep (hunk 4 above)
Comment (MrMCake) [reply arm/Microsoft.Subscription/aliases/deploy.bicep]: done

File: arm/Microsoft.Subscription/aliases/deploy.bicep (hunk 5 above)
Comment (MrMCake) [reply arm/Microsoft.Subscription/aliases/deploy.bicep:84]: Added

**Comment by AlexanderSehr** [comment]
Closing for now and may be re-implemented in AVM if need be.

**Pull request closed without merging**